Use more std::span<uint8_t> in IPC::Decoder

ddkilzer opened the pull request:
#### 1e7b936a641ccf787c0362a26d984a65463429ca
<pre>
Use more std::span&lt;uint8_t&gt; in IPC::Decoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=261228">https://bugs.webkit.org/show_bug.cgi?id=261228</a>
&lt;rdar://115075685&gt;

Reviewed by Chris Dumez.

Switch IPC::Decoder to use DataReference (std::span&lt;const uint8_t&gt;).

Other changes of note:
- Add support for std::span&lt;uint8_t&gt; when creating JSC::ArrayBuffer.
- Add a utility method for std::span to &lt;wtf/Algorithms.h&gt;.
- Introduce MessageName::Invalid and initialize
  IPC::Decoder::m_message_name to that value.
- IPC::Decoder objects now fit in 64 bytes.
- IPC::Decoder now calls m_bufferDeallocator when marked invalid,
  and clears m_buffer.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBuffer::create): Add.
(JSC::ArrayBuffer::tryCreate): Add.
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
(JSC::ArrayBuffer::create): Add.
(JSC::ArrayBuffer::tryCreate): Add.

* Source/WTF/wtf/Algorithms.h:
(WTF::spanReinterpretCast): Add.

* Source/WebKit/NetworkProcess/Notifications/Cocoa/WebPushDaemonConnectionCocoa.mm:
(WebKit::WebPushD::Connection::performSendWithAsyncReplyWithoutUsingIPCConnection const):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::copyBuffer):
(IPC::Decoder::create):
- Remove duplicate code for one overloaded method.
(IPC::Decoder::Decoder):
- Delete default constructor.
(IPC::Decoder::~Decoder):
(IPC::Decoder::unwrapForTesting):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::buffer const): Switch to return DataReference.
(IPC::Decoder::currentBufferPosition const): Delete.
- Rename to currentBufferOffset() since this returns an offset from the
  start of the buffer.
(IPC::Decoder::currentBufferOffset const):
(IPC::Decoder::length const): Delete.
(IPC::Decoder::isValid const):
(IPC::Decoder::markInvalid):
- Now calls m_bufferDeallocator and clear m_buffer on invalidation.
(IPC::alignedBufferIsLargeEnoughToContain):
- Simplify logic because we are no longer using two pointer addresses.
(IPC::Decoder::decodeSpan):
* Source/WebKit/Platform/IPC/JSIPCBinding.cpp:
(IPC::putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined):
* Source/WebKit/Platform/IPC/JSIPCBinding.h:
(IPC::putJSValueForDecodeArgumentInArray):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::dispatchStreamMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::dispatchOutOfStreamMessage):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::createMessageDecoder):
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::processMessage):
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
(IPC::Connection::readEventHandler):
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):
- Add MessageName::Invalid.
* Source/WebKit/Scripts/webkit/tests/MessageNames.h: Ditto.
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::defaultTestDriver):
(WebKit::sendTestMessage):
* Source/WebKit/UIProcess/LegacySessionStateCodingNone.cpp:
(WebKit::decodeLegacySessionState):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSMessageListener::didReceiveMessage):
(WebKit::IPCTestingAPI::JSMessageListener::willSendMessage):
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::connectionEventHandler):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::performSendWithAsyncReplyWithoutUsingIPCConnection const):
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::Encoder&gt;::createDecoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest&lt;IPC::StreamConnectionEncoder&gt;::createDecoder const):
(TestWebKitAPI::TYPED_TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::copyViaEncoder):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::WebPushXPCConnectionMessageSender::performSendWithAsyncReplyWithoutUsingIPCConnection const):

Canonical link: <a href="https://commits.webkit.org/267956@main">https://commits.webkit.org/267956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34264b4a9698800d9a7bc1e64d775836850bf31e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18164 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18497 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19067 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20006 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18362 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21794 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18656 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18616 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15816 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20883 "Built successfully") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/18322 "Found 1 webkitpy python3 test failure: webkitpy.w3c.test_converter_unittest.W3CTestConverterTest.test_convert_prefixed_properties") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15840 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16569 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23085 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15741 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16859 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16740 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20968 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/17450 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17311 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14666 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21520 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16399 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5271 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20763 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/22753 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2228 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17158 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5118 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -208,6 +208,14 @@ Ref<ArrayBuffer> ArrayBuffer::create(const void* source, size_t byteLength)
     return buffer.releaseNonNull();
 }
 
+Ref<ArrayBuffer> ArrayBuffer::create(std::span<uint8_t>&& span)
```
We usually pass std::span by value, it is cheap to copy.

ddkilzer replied:
I will change all of these, but there is pre-existing code in WebKit that uses r-value references for spans.

The thing I like about using a reference is that it can be explicitly cleared out so that if someone tries to use it afterwards, it will produce an error, rather than working "by accident".

cdumez replied:
Because std::span<> doesn't own the memory, it is completely fine to keep using the span after calling this create() function though.

cdumez commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -208,6 +208,14 @@ Ref<ArrayBuffer> ArrayBuffer::create(const void* source, size_t byteLength)
 
+Ref<ArrayBuffer> ArrayBuffer::create(std::span<uint8_t>&& span)
+{
+    auto buffer = tryCreate(std::forward<std::span<uint8_t>>(span));
```
no std::forward.

cdumez commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -268,6 +276,12 @@ RefPtr<ArrayBuffer> ArrayBuffer::tryCreate(const void* source, size_t byteLength
     return createInternal(WTFMove(contents), source, byteLength);
 }
 
+RefPtr<ArrayBuffer> ArrayBuffer::tryCreate(std::span<uint8_t>&& span)
```
pass by value.

cdumez commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -268,6 +276,12 @@ RefPtr<ArrayBuffer> ArrayBuffer::tryCreate(const void* source, size_t byteLength
 
+RefPtr<ArrayBuffer> ArrayBuffer::tryCreate(std::span<uint8_t>&& span)
+{
+    auto s = std::exchange(span, { });
```
This whole line is not needed.

ddkilzer replied:
> The thing I like about using a reference is that it can be explicitly cleared out so that if someone tries to use it afterwards, it will produce an error, rather than working "by accident".

cdumez commented on Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
```diff
@@ -268,6 +276,12 @@ RefPtr<ArrayBuffer> ArrayBuffer::tryCreate(const void* source, size_t byteLength
+RefPtr<ArrayBuffer> ArrayBuffer::tryCreate(std::span<uint8_t>&& span)
+{
+    auto s = std::exchange(span, { });
+    return tryCreate(s.data(), s.size_bytes());
```
can use `span` directly here.

cdumez commented on Source/WTF/wtf/Algorithms.h:
```diff
@@ -54,5 +58,40 @@ bool allOf(ContainerType&& container, AllOfFunction allOfFunction)
+template<typename T, typename U>
+std::span<T> spanCast(std::span<U> span)
+{
+    RELEASE_ASSERT(!(span.size_bytes() % sizeof(T)));
```
Don't we want to static_assert() that T and U have the same size?

ddkilzer replied:
> Don't we want to static_assert() that T and U have the same size?

No, that's the whole point of this method.  To cast from a `uint8_t` span to a span of a different type.  We just want to make sure that we don't have any leftover bytes.

Unless you would like a size (count) of `T` objects passed in to make sure that converting from `U` to `T` also has the expected number of `T` objects as well.

This is used in the `IPC::Decoder` class.

cdumez commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -27,58 +27,53 @@
 namespace IPC {
 
-static const uint8_t* copyBuffer(const uint8_t* buffer, size_t bufferSize)
+static DataReference copyBuffer(DataReference buffer)
```
I find it a little weird to be storing newly malloc-d memory in a std::span. I'd normally not think of a span as something I need to explicitly free() when I am done.

Maybe that's ok but a little confusing?

ddkilzer replied:
Okay, I guess that would get rid of the need for `memcpySpan()` below.

cdumez commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -49,13 +51,14 @@ template<typename, typename, typename> struct HasModernDecoder;
-    static std::unique_ptr<Decoder> create(const uint8_t* buffer, size_t bufferSize, BufferDeallocator&&, Vector<Attachment>&&);
-    Decoder(const uint8_t* stream, size_t streamSize, uint64_t destinationID);
+    static std::unique_ptr<Decoder> create(DataReference buffer, Vector<Attachment>&&);
+    using BufferDeallocator = Function<void(DataReference&&)>;
```
No need for `&&`.

cdumez commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -49,13 +51,14 @@ template<typename, typename, typename> struct HasModernDecoder;
+    static std::unique_ptr<Decoder> create(DataReference buffer, Vector<Attachment>&&);
+    using BufferDeallocator = Function<void(DataReference&&)>;
+    static std::unique_ptr<Decoder> create(DataReference&& buffer, BufferDeallocator&&, Vector<Attachment>&&);
+    Decoder(DataReference&& stream, uint64_t destinationID);
```
No need for `&&`

cdumez commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -145,33 +152,32 @@ class Decoder {
 
 private:
-    Decoder(const uint8_t* buffer, size_t bufferSize, BufferDeallocator&&, Vector<Attachment>&&);
+    Decoder(DataReference&& buffer, BufferDeallocator&&, Vector<Attachment>&&);
```
No need for `&&`.

cdumez commented on Source/WTF/wtf/Algorithms.h:
```diff
@@ -54,5 +58,40 @@ bool allOf(ContainerType&& container, AllOfFunction allOfFunction)
+}
+
+template<typename T, typename U>
+std::span<T> spanCast(std::span<U> span)
```
spanReinterpretCast() given what it actually does?

ddkilzer replied:
Sounds good.

cdumez commented on Source/WebKit/Platform/IPC/JSIPCBinding.cpp:
```diff
@@ -194,12 +194,12 @@ JSC::JSValue jsValueForDecodedArgumentValue(JSC::JSGlobalObject* globalObject, W
 }
 
-bool putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined(JSC::JSGlobalObject* globalObject, JSC::JSArray* array, unsigned index, JSC::JSValue value, const uint8_t* buffer, size_t length)
+bool putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined(JSC::JSGlobalObject* globalObject, JSC::JSArray* array, unsigned index, JSC::JSValue value, DataReference&& buffer)
```
No need for `&&`

cdumez commented on Source/WebKit/Platform/IPC/JSIPCBinding.cpp:
```diff
@@ -194,12 +194,12 @@ JSC::JSValue jsValueForDecodedArgumentValue(JSC::JSGlobalObject* globalObject, W
 
     if (value.isUndefined()) {
-        auto arrayBuffer = JSC::ArrayBuffer::create(buffer, length);
+        auto arrayBuffer = JSC::ArrayBuffer::create(std::forward<DataReference>(buffer));
```
no std::forward needed.

cdumez commented on Source/WebKit/Platform/IPC/JSIPCBinding.h:
```diff
@@ -122,7 +122,7 @@ JSC::JSValue jsValueForDecodedArgumentValue(JSC::JSGlobalObject* globalObject, O
 }
 
-bool putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined(JSC::JSGlobalObject*, JSC::JSArray*, unsigned index, JSC::JSValue, const uint8_t* buffer, size_t length);
+bool putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined(JSC::JSGlobalObject*, JSC::JSArray*, unsigned index, JSC::JSValue, DataReference&& buffer);
```
No `&&` needed.

cdumez commented on Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
```diff
@@ -450,9 +450,9 @@ static void clearNoSenderNotifications(mach_port_t port)
         descriptor->out_of_line.deallocate = false; // We are taking ownership of the memory.
 
-        return Decoder::create(messageBody, messageBodySize, [](const uint8_t* buffer, size_t length) {
+        return Decoder::create({ messageBody, messageBodySize }, [](DataReference&& buffer) {
```
No `&&` needed.

cdumez commented on Source/WebKit/Shared/IPCTester.cpp:
```diff
@@ -72,14 +72,15 @@ static void defaultTestDriver(WKMessageTestSendMessageFunc sendMessageFunc, void
         return 1;
     BinarySemaphore semaphore;
-    auto decoder = IPC::Decoder::create(data, size, [&semaphore] (const uint8_t*, size_t) { semaphore.signal(); }, { }); // NOLINT
+    auto decoder = IPC::Decoder::create({ data, size }, [&semaphore] (IPC::DataReference&&) { semaphore.signal(); }, { }); // NOLINT
```
No `&&` needed.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -539,7 +539,7 @@ bool WebProcessProxy::shouldSendPendingMessage(const PendingMessage& message)
     if (message.encoder->messageName() == IPC::MessageName::WebPage_LoadRequestWaitingForProcessLaunch) {
         auto buffer = message.encoder->buffer();
         auto bufferSize = message.encoder->bufferSize();
```
It may be nice for `message.encoder` to return us the buffer as a std::span directly.

ddkilzer replied:
This patch already snowballed a bit.  If I change that, I suspect this patch will get even bigger.

I think it will be obvious when changing `IPC::Encoder` to use `std::span` to make this change then.

cdumez replied:
All I suggested was introducing a new member function that returns a span, not modify any existing member function. Therefore, I don't understand the snowball risk. It only makes the code you change nicer.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -559,7 +559,7 @@ bool WebProcessProxy::shouldSendPendingMessage(const PendingMessage& message)
     } else if (message.encoder->messageName() == IPC::MessageName::WebPage_GoToBackForwardItemWaitingForProcessLaunch) {
         auto buffer = message.encoder->buffer();
         auto bufferSize = message.encoder->bufferSize();
```
ditto.

cdumez commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2864,7 +2864,7 @@ void JSMessageListener::didReceiveMessage(const IPC::Decoder& decoder)
     JSC::JSLockHolder lock(globalObject->vm());
 
-    auto mutableDecoder = IPC::Decoder::create(decoder.buffer(), decoder.length(), { });
+    auto mutableDecoder = IPC::Decoder::create({ decoder.buffer(), decoder.length() }, { });
```
Maybe `decoder` could return the buffer as a span directly.

ddkilzer replied:
It actually does.  I missed this somehow, and the bots caught the build error.

cdumez commented on Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
```diff
@@ -2881,7 +2881,7 @@ void JSMessageListener::willSendMessage(const IPC::Encoder& encoder, OptionSet<I
     JSC::JSLockHolder lock(globalObject->vm());
 
-    auto decoder = IPC::Decoder::create(encoder.buffer(), encoder.bufferSize(), { });
+    auto decoder = IPC::Decoder::create({ encoder.buffer(), encoder.bufferSize() }, { });
```
ditto.

cdumez commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -27,58 +27,52 @@
 namespace IPC {
 
-static const uint8_t* copyBuffer(const uint8_t* buffer, size_t bufferSize)
+static const uint8_t* copyBuffer(DataReference buffer)
```
Not new to this patch so you don't have to fix before landing but I don't think this should return a `const` pointer since it points to newly malloc'd memory that  the caller needs to free.

ddkilzer replied:
Will fix.

cdumez commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -27,58 +27,52 @@
-        return nullptr;
-    }
-    return Decoder::create(copyBuffer(buffer, bufferSize), bufferSize, [](const uint8_t* ptr, size_t) { fastFree(const_cast<uint8_t*>(ptr)); }, WTFMove(attachments)); // NOLINT
+    return Decoder::create({ copyBuffer(buffer), buffer.size() }, [](DataReference buffer) { fastFree(const_cast<uint8_t*>(buffer.data())); }, WTFMove(attachments)); // NOLINT
```
We shouldn't need a const_cast here.

ddkilzer replied:
The compiler says we do need it:

```
OpenSource/Source/WebKit/Platform/IPC/Decoder.cpp:52:94: error: no matching function for call to 'fastFree'
    return Decoder::create({ copyBuffer(buffer), buffer.size() }, [](DataReference buffer) { fastFree(buffer.data()); }, WTFMove(attachments)); // NOLINT
                                                                                             ^~~~~~~~
In file included from WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/PlatformUnifiedSource3.cpp:1:
In file included from OpenSource/Source/WebKit/Platform/IPC/DaemonEncoder.cpp:26:
In file included from OpenSource/Source/WebKit/config.h:42:
WebKitBuild/Debug/usr/local/include/wtf/FastMalloc.h:103:25: note: candidate function not viable: no known conversion from 'std::span<const unsigned char, 18446744073709551615>::pointer' (aka 'const unsigned char *') to 'void *' for 1st argument
WTF_EXPORT_PRIVATE void fastFree(void*);
                        ^
1 error generated.
```

Or are you thinking we should change the definition of `DataReference` to `std::span<uint8_t>`?

ddkilzer replied:
Chris and I chatted in person, and changing `DataReference` to `std::span<uint8_t>` is a bigger change than affects more code than just this patch, and may have other implications, so I think it would be best (if we decide to do it) to land as a separate patch.

cdumez commented on Source/WebKit/Platform/IPC/Decoder.cpp:
```diff
@@ -27,58 +27,52 @@
 }
 
-std::unique_ptr<Decoder> Decoder::create(const uint8_t* buffer, size_t bufferSize, Vector<Attachment>&& attachments)
+std::unique_ptr<Decoder> Decoder::create(DataReference buffer, Vector<Attachment>&& attachments)
```
Not a regression from this patch but it seems wrong to use a const pointer (and now a `span<const uint8_t>`) to represent memory we need to free.

ddkilzer replied:
Well, sometimes we free it (if `BufferDeallocator` is set) and sometimes we don't free it (if `BufferDeallocator` is `nullptr`).

Are you suggesting changing `DataReference` to be `std::span<uint8_t>` everywhere, or just change one `Decoder` constructor and one `Decoder::create()` method to use `std::span<uint8_t>` in place of `DataReference`?

Or do we need a `MutableDataReference` that's `std::span<uint8_t>` to differentiate `DataReference` that's `std::span<const uint8_t>`?

I tried this, and it got messy quickly, so I think this should probably be done in a separate patch.

ddkilzer replied:
See above.

cdumez reviewed (approved):
LGTM

ddkilzer commented on Source/WebKit/Shared/IPCTester.cpp:
```diff
@@ -72,14 +72,15 @@ static void defaultTestDriver(WKMessageTestSendMessageFunc sendMessageFunc, void
 
 static int sendTestMessage(const uint8_t* data, size_t size, void* context)
 {
+    // FIXME: Switch to IPC::DataReference for first two arguments.
```
I have a fix for this that I will include shortly.

Commit:
Use more std::span<uint8_t> in IPC::Decoder
https://bugs.webkit.org/show_bug.cgi?id=261228
<rdar://115075685>

Reviewed by Chris Dumez.

Switch IPC::Decoder to use DataReference (std::span<const uint8_t>).

Other changes of note:
- Add support for std::span<uint8_t> when creating JSC::ArrayBuffer.
- Add a utility method for std::span to <wtf/Algorithms.h>.
- Introduce MessageName::Invalid and initialize
  IPC::Decoder::m_message_name to that value.
- IPC::Decoder objects now fit in 64 bytes.
- IPC::Decoder now calls m_bufferDeallocator when marked invalid,
  and clears m_buffer.

* Source/JavaScriptCore/runtime/ArrayBuffer.cpp:
(JSC::ArrayBuffer::create): Add.
(JSC::ArrayBuffer::tryCreate): Add.
* Source/JavaScriptCore/runtime/ArrayBuffer.h:
(JSC::ArrayBuffer::create): Add.
(JSC::ArrayBuffer::tryCreate): Add.

* Source/WTF/wtf/Algorithms.h:
(WTF::spanReinterpretCast): Add.

* Source/WebKit/NetworkProcess/Notifications/Cocoa/WebPushDaemonConnectionCocoa.mm:
(WebKit::WebPushD::Connection::performSendWithAsyncReplyWithoutUsingIPCConnection const):
* Source/WebKit/Platform/IPC/Decoder.cpp:
(IPC::copyBuffer):
(IPC::Decoder::create):
- Remove duplicate code for one overloaded method.
(IPC::Decoder::Decoder):
- Delete default constructor.
(IPC::Decoder::~Decoder):
(IPC::Decoder::unwrapForTesting):
* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::Decoder::buffer const): Switch to return DataReference.
(IPC::Decoder::currentBufferPosition const): Delete.
- Rename to currentBufferOffset() since this returns an offset from the
  start of the buffer.
(IPC::Decoder::currentBufferOffset const):
(IPC::Decoder::length const): Delete.
(IPC::Decoder::isValid const):
(IPC::Decoder::markInvalid):
- Now calls m_bufferDeallocator and clear m_buffer on invalidation.
(IPC::alignedBufferIsLargeEnoughToContain):
- Simplify logic because we are no longer using two pointer addresses.
(IPC::Decoder::decodeSpan):
* Source/WebKit/Platform/IPC/JSIPCBinding.cpp:
(IPC::putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined):
* Source/WebKit/Platform/IPC/JSIPCBinding.h:
(IPC::putJSValueForDecodeArgumentInArray):
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
* Source/WebKit/Platform/IPC/StreamServerConnection.cpp:
(IPC::StreamServerConnection::dispatchStreamMessages):
(IPC::StreamServerConnection::processSetStreamDestinationID):
(IPC::StreamServerConnection::dispatchStreamMessage):
(IPC::StreamServerConnection::dispatchOutOfStreamMessage):
* Source/WebKit/Platform/IPC/cocoa/ConnectionCocoa.mm:
(IPC::createMessageDecoder):
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp:
(IPC::Connection::processMessage):
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
(IPC::Connection::readEventHandler):
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):
- Add MessageName::Invalid.
* Source/WebKit/Scripts/webkit/tests/MessageNames.h: Ditto.
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::defaultTestDriver):
(WebKit::sendTestMessage):
* Source/WebKit/UIProcess/LegacySessionStateCodingNone.cpp:
(WebKit::decodeLegacySessionState):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::shouldSendPendingMessage):
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::JSMessageListener::didReceiveMessage):
(WebKit::IPCTestingAPI::JSMessageListener::willSendMessage):
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::connectionEventHandler):
* Source/WebKit/webpushd/webpushtool/WebPushToolConnection.mm:
(WebPushTool::Connection::performSendWithAsyncReplyWithoutUsingIPCConnection const):
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::Encoder>::createDecoder const):
(TestWebKitAPI::ArgumentCoderEncoderDecoderTest<IPC::StreamConnectionEncoder>::createDecoder const):
(TestWebKitAPI::TYPED_TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::copyViaEncoder):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::WebPushXPCConnectionMessageSender::performSendWithAsyncReplyWithoutUsingIPCConnection const):

Canonical link: https://commits.webkit.org/267956@main
